File: 1.1 Merge Sort and Input/merge_sort.cc
// Merge Sort (Integers) and Input
// Bess L. Walker
// 1-31-2012
// Hurrah, new and delete instead of calloc and free

#include <iostream>

using namespace std;

//pre-declarations
int* merge_sort(int* unsorted, int length);
int* merge(int* left, int left_length, int* right, int right_length);
void pprint_intarray(int intarray [], int length);

int main()
{
	int permut_1 [] = {0, 1, 2, 3, 4, 5};
	int permut_1_len = 6;
	pprint_intarray(permut_1, permut_1_len);
	cout << endl;
	pprint_intarray(merge_sort(permut_1, permut_1_len), permut_1_len);
	
	cout << endl;
	
	int permut_2 [] = {5, 4, 3, 2, 1, 0};
	int permut_2_len = 6;
	pprint_intarray(permut_2, permut_2_len);
	cout << endl;
	pprint_intarray(merge_sort(permut_2, permut_2_len), permut_2_len);
	
	cout << endl;
	
	int permut_3 [] = {3, 1, 31, 10, 1, 5};
	int permut_3_len = 6;
	pprint_intarray(permut_3, permut_3_len);
	cout << endl;
	pprint_intarray(merge_sort(permut_3, permut_3_len), permut_3_len);
	
	cout << endl;
	
	return 0;
}

int*  merge_sort(int* unsorted, int length)
{
	if (unsorted == NULL || length == 0)
	{return NULL;}
	else if (length == 1)
	{return unsorted;}
	
	int* sorted = NULL;
	int split_index = length / 2;  // integer division on purpose
	
	int* left = merge_sort(unsorted, split_index);
	int* right = merge_sort(unsorted + split_index, length - split_index);
	
	sorted = merge(left, split_index, right, length - split_index);
	
	// if left and/or right came from a merge, they need to be deleted
	// that is, if they have length > 1
	if (split_index > 1) { delete left; }
	if (length - split_index > 1) { delete right; }
	
	return sorted;
}

int* merge(int* left, int left_length, int* right, int right_length)
{
	int* merged = new int [left_length + right_length];
	int merged_index = 0;
	int left_index = 0;
	int right_index = 0;
	
	while (left_index < left_length && right_index < right_length)
	{
		if (left[left_index] <= right[right_index])
		{
			merged[merged_index] = left[left_index];
			merged_index++;
			left_index++;
		}
		else
		{
			merged[merged_index] = right[right_index];
			merged_index++;
			right_index++;
		}
	}
	
	while (left_index < left_length)
	{
		merged[merged_index] = left[left_index];
		merged_index++;
		left_index++;
	}
	
	while (right_index < right_length)
	{
		merged[merged_index] = right[right_index];
		merged_index++;
		right_index++;
	}
	
	return merged;
}

void pprint_intarray(int intarray [], int length)
{
	cout << "[";
	if (length >= 1) { cout << " " << intarray[0]; }
	for(int ii = 1; ii < length; ii++)
	{ cout << ", " << intarray[ii]; }
	cout << "]";
}

File: 1.1 Merge Sort and Input/merge_sort.py
# Merge Sort (Integers) and Input
# Bess L. Walker
# 11-17-11
# Python first this time; let's renew familiarity with the algorithm while 
# not also trying to remember the language...

def prompt():
	int_list = []
	bad_list = []

	string = raw_input("Enter a list of integers separated by commas: ")
	
	# Tokenize
	tokens = string.strip()
	tokens = string.split(",")
	
	# Legal?  Add to integer list.	Illegal?  Add to bad input list.
	for token in tokens:
		token = token.strip()
		try:
			integer = int(token)
			int_list.append(integer)
		except:
			bad_list.append(token)
			
	print("Your list of integers is: " + str(int_list))
	if len(bad_list) > 0:
		print("Your bad input was filtered out: " + str(bad_list))
		
	return int_list
				
def merge_sort(unsorted_list):
	length = len(unsorted_list)
	
	if length in [0, 1]:
		return unsorted_list

	index_split = length / 2		# integer division is on purpose

	left = merge_sort(unsorted_list[:index_split])
	right = merge_sort(unsorted_list[index_split:])
	
	sorted_list = merge(left, right)
	
	return sorted_list

# It's not clever, but it will work.
# I'd love to write something more in-place, but I can't seem to do it
# without blowing merge() up to worst-case n^2.
# Oh well, one of the well-known issues with merge sort is that it is not in-place.
def merge(left, right):
	left_index = 0
	right_index = 0

	merged_list = []
	
	while left_index < len(left) and right_index < len(right):
		if left[left_index] <= right[right_index]:
			merged_list.append(left[left_index])
			left_index += 1
		else:
			merged_list.append(right[right_index])
			right_index += 1
		
	if left_index < len(left):
		merged_list += left[left_index:]
		
	if right_index < len(right):
		merged_list += right[right_index:]
			
	return merged_list
	



# At least in this implementation, sorted input doesn't take less time
# than anything else.  Oh well.
permut_1 = [0, 1, 2, 3, 4, 5]
print permut_1
print merge_sort(permut_1)

print

permut_2 = [5, 4, 3, 2, 1, 0]
print permut_2
print merge_sort(permut_2)

print

permut_3 = [5, 11, 17, 11, 5, 3]
print permut_3
print merge_sort(permut_3)

print

unsorted = prompt()
print
print unsorted
print merge_sort(unsorted)
